### PR TITLE
Change shutdown stalled method to also check/include the main/pipeline thread

### PR DESCRIPTION
**Pull request by edmocosta**
## Release notes

[rn:skip]

## What does this PR do?

This PR changes the [java_pipeline.rb](https://github.com/elastic/logstash/pull/15056/files#diff-f1d310cd201ca9fcfad2f54ecdfd588cc476eb5ecee6939e330547eed0476208) to include the pipeline/main thread on the stalling threads info list, that way, Logsatsh can provide users more helpful information when the stalling thread is the main one.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have made corresponding change to the default configuration files (and/or docker env variables)
- [x] I have added tests that prove my fix is effective or that my feature works

## Related issues

- Closes https://github.com/elastic/logstash/issues/14977

## Logs

**Before:**

```
[2023-03-21T14:34:36,415][WARN ][logstash.runner ] SIGTERM received. Shutting down.
[2023-03-21T14:34:37,093][INFO ][filewatch.observingread ] QUIT - closing all files and shutting down.
[2023-03-21T14:34:37,380][INFO ][logstash.inputs.elasticsearch][my-pipeline-2][11e442e608e1c8343861f241261f13bea2b7ece0cb11ea5e455421067d951f5f] Slice complete {:slice_id=>1, :slices=>2}
[2023-03-21T14:34:39,537][INFO ][logstash.javapipeline ][my-pipeline-1] Pipeline terminated {"pipeline.id"=>"my-pipeline-1"}
[2023-03-21T14:34:40,447][INFO ][logstash.pipelinesregistry] Removed pipeline from registry successfully {:pipeline_id=>:"my-pipeline-1"}
[2023-03-21T14:34:43,341][WARN ][org.logstash.execution.ShutdownWatcherExt] {"inflight_count"=>0, "stalling_threads_info"=>{}}
[2023-03-21T14:34:43,368][ERROR][org.logstash.execution.ShutdownWatcherExt] The shutdown process appears to be stalled due to busy or blocked plugins. Check the logs for more information.
[2023-03-21T14:34:43,429][INFO ][logstash.javapipeline ][my-pipeline-2] Pipeline terminated {"pipeline.id"=>"my-pipeline-2"}
[2023-03-21T14:34:44,407][INFO ][logstash.pipelinesregistry] Removed pipeline from registry successfully {:pipeline_id=>:"my-pipeline-2"}
```

**After:**
```
[2023-06-06T17:21:48,728][WARN ][logstash.runner          ] SIGTERM received. Shutting down.
[2023-06-06T17:21:48,729][DEBUG][logstash.agent           ] Shutting down all pipelines {:pipelines_count=>1}
[2023-06-06T17:21:48,730][DEBUG][logstash.agent           ] Converging pipelines state {:actions_count=>1}
[2023-06-06T17:22:08,808][DEBUG][logstash.instrument.periodicpoller.cgroup] One or more required cgroup files or directories not found: /proc/self/cgroup, /sys/fs/cgroup/cpuacct, /sys/fs/cgroup/cpu
[2023-06-06T17:22:08,809][DEBUG][logstash.instrument.periodicpoller.jvm] collector name {:name=>"G1 Young Generation"}
[2023-06-06T17:22:08,810][DEBUG][logstash.instrument.periodicpoller.jvm] collector name {:name=>"G1 Old Generation"}
[2023-06-06T17:22:08,974][WARN ][org.logstash.execution.ShutdownWatcherExt] {"inflight_count"=>10, "stalling_threads_info"=>{["LogStash::Filters::Sleep", {"id"=>"sleep", "time"=>30}]=>[{"thread_id"=>43, "name"=>"[stalling]>worker0", "current_call"=>"[...]/vendor/bundle/jruby/2.6.0/gems/logstash-filter-sleep-3.0.7/lib/logstash/filters/sleep.rb:105:in `sleep'"}, {"thread_id"=>44, "name"=>"[stalling]>worker1", "current_call"=>"[...]/vendor/bundle/jruby/2.6.0/gems/logstash-filter-sleep-3.0.7/lib/logstash/filters/sleep.rb:105:in `sleep'"}, {"thread_id"=>45, 
```

### CODE REVIEW

**Comment by edmocosta** [comment]
Jenkins test this please

**Review by jsvd — CHANGES_REQUESTED** [review]
so here we're addressing the fact that if both counts and threads are empty then we don't want to mark the pipeline as stalled.

Separately, it is reasonable to mark the pipeline as stalled. Either:
a) there are no inflight events but there are still threads executing. OR:
b) there are inflight events but no stalling threads

Both of these should be rare, but they are real justifications for the pipeline not shutting down.

So I'd suggest changing the logic to mark the pipeline as stalled iff:
1. the number of inflight events is not decreasing and > 0, or
2. number of stalling threads is > 0

Either if these is enough to mark as stalled, or both.

Writing tests for these scenarios will help:
a) 5 reports with 0 inflight events + no stalled threads => not stalled
b) 5 reports with 0 inflight events + stalled threads => stalled
c) 5 reports with > 0 inflight events + no stalled threads => stalled
c) 5 reports with > 0 inflight events + stalled threads => stalled

File: logstash-core/src/main/java/org/logstash/execution/ShutdownWatcherExt.java [hunk 1]
```diff
@@ -20,22 +20,21 @@
...
 import org.jruby.anno.JRubyMethod;
 import org.jruby.runtime.ThreadContext;
 import org.jruby.runtime.builtin.IRubyObject;
-import org.logstash.RubyUtil;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
```
Comment (jsvd) [inline logstash-core/src/main/java/org/logstash/execution/ShutdownWatcherExt.java:37]: If possible I'd prefer not to have changes in the imports unless they're necessary.

File: logstash-core/src/main/java/org/logstash/execution/ShutdownWatcherExt.java [hunk 2]
```diff
@@ -105,29 +104,35 @@ public IRubyObject shutdownStalled(final ThreadContext context) {
...
-                    break;
-                }
+
+        final IRubyObject[] stallingThreads = reports.stream()
+                .map(obj -> obj.callMethod(context, "stalling_threads"))
+                .toArray(IRubyObject[]::new);
+
+        if (inflightCounts.length == 0 && stallingThreads.length == 0) {
+            return context.fals;
+        }
```
Comment (jsvd) [inline logstash-core/src/main/java/org/logstash/execution/ShutdownWatcherExt.java:124]: Given that these are derived from the reports array which is guaranteed to be of size `reportEvery`, these arrays should also be of size `reportEvery` (5)

File: logstash-core/src/main/java/org/logstash/execution/ShutdownWatcherExt.java [hunk 3]
```diff
@@ -105,29 +104,35 @@ public IRubyObject shutdownStalled(final ThreadContext context) {
...
+                return context.fals;
             }
-            return stalled ? context.tru : context.fals;
         }
-        return context.fals;
+
+        final int lastInflightCount = inflightCounts.length > 0 ? inflightCounts[inflightCounts.length - 1] : 0;
+        return lastInflightCount > 0 || stallingThreads.length > 0
+                ? context.tru
+                : context.fals;
```
Comment (jsvd) [inline logstash-core/src/main/java/org/logstash/execution/ShutdownWatcherExt.java:135]: if all inflight counts are 0, then we look at `stallingthreads`, but that will always have a length of 5, so it's always true. We'd have to look at the last element of that array instead

**Review by jsvd — CHANGES_REQUESTED** [review]
we should have a couple of tests for when inflight count is always 0 on all 5 reports. 
also added a few comments about the logic

**Review by jsvd — CHANGES_REQUESTED** [review]
the test failure is legit since it needs to now address the existance of a @thread, which in the test is likely to be nil

**Review by jsvd — APPROVED** [review]
minor nitpick, otherwise LGTM

**Comment by essodjolo** [comment]
Hi @jsvd,

I have a user waiting for this fix. They reported that the issue is still happening in `8.8.2`. Can you confirm which version will embed this fix please?
Thanks.

Ps: I couldn't see any label pointing to the target version.

**Comment by jsvd** [comment]
@logstashmachine backport 7.17

**Comment by jsvd** [comment]
@logstashmachine backport 8.9

**Comment by jsvd** [comment]
@logstashmachine backport 8.9
